### ISSUE TITLE
Type narrowing fails during structural pattern matching on NotRequired TypedDict fields

### ISSUE THREAD

**Issue by sirrus233** [microsoft/pyright#5056]
**Describe the bug**
When using a mapping pattern for structural pattern matching on a `TypedDict` with a `NotRequired` field (from `typing_extensions`, pyright does not understand that if the pattern is matched then the key must exist.

**To Reproduce**
```
class Foo(TypedDict):
    value: NotRequired[int]


def increment_if_exists(foo: Foo) -> None:
    match foo:
        case {"value": _}:
            foo["value"] += 1  # error: reportTypedDictNotRequiredAccess
        case _:
            pass
```
This gives the error:
```
error: Could not access item in TypedDict
    "value" is not a required key in "Foo", so access may result in runtime exception (reportTypedDictNotRequiredAccess)
```

**Expected behavior**
Accessing the `value` field in the example above should be safe, because the pattern has been matched.

**VS Code extension or command-line**
Issue occurs in VSCode extension and cmdline Pyright
`pyright ==  1.1.304`
`typing-extensions == 4.5.0`

**Comment by erictraut** [comment]
Yes, pyright doesn't currently support this form of type narrowing for mapping patterns. I think it's a reasonable request given that it does support type narrowing for the traditional `"key" in dict` type guard. You can observe this in action by adding the statement `assert "value" in foo` to your above code.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.307, which I just published. It will also be included in a future release of pylance.
